File: CCAMA.cpp
#include <iostream>
#include <iomanip>
#include <vector>
#include <Eigen/Dense>
#include <Eigen/Eigenvalues>
#include <Eigen/SVD>
#include <Eigen/LU>
#include <math.h>
#include "Lyap.h"
#include "Options.h"
#include "Output.h"

using namespace std;
using namespace Eigen;
using Eigen::MatrixXd;
using Eigen::VectorXd;

Output ccama(MatrixXd A, MatrixXd C, MatrixXd E, MatrixXd G,
	int gamma, int n, int m, Options options)
{
	int dispW = 20;

	// declarations needed for output
	Output output;
	MatrixXd Y1new;
	MatrixXd Y2new;
	MatrixXd Xnew;
	MatrixXd Znew;

	// define identity matricies of size m and n for use throughout program
	MatrixXd I_m(m, m);
	I_m = MatrixXd::Identity(I_m.rows(), I_m.cols());

	MatrixXd I_n(n, n);
	I_n = MatrixXd::Identity(I_n.rows(), I_n.cols());

	if (!options.useOptions)
	{
		options.rho = 10;
		options.epsPrim = 1.e-5;
		options.epsDual = 1.e-4;
		options.maxIter = 1.e5;
		MatrixXd Xinit = Lyap(A, I_m);
		options.xInit = Xinit;
		options.zInit = I_m;
		MatrixXd Y1init = Lyap(A.transpose(), -Xinit);
		double tempNorm = Y1init.norm();
		options.yOneInit = (gamma * Y1init) / tempNorm;
		options.yTwoInit = I_n;
		options.method = 1;
	}

	// default options parameters if no options are given


	// Data processing
	double rho0 = options.rho;
	double eps_prim = options.epsPrim;
	double eps_dual = options.epsDual;
	int MaxIter = options.maxIter;
	MatrixXd X = options.xInit;
	MatrixXd Z = options.zInit;
	MatrixXd Y1 = options.yOneInit;
	MatrixXd Y2 = options.yTwoInit;


	// AMA implementation
	if (options.method == 1)
	{
		MatrixXd tempEigenSolver = A.transpose() * Y1 + Y1 * A + C.transpose() * E.cwiseProduct(Y2) * C;
		VectorXd compEigLadY = tempEigenSolver.eigenvalues();
		VectorXd eigLadY = compEigLadY.real();
		ArrayXXd arrayTemp = eigLadY.array().log();
		double logdetLadY = arrayTemp.matrix().sum();
		MatrixXd traceMatrix = G * Y2;
		complex<double> dualY = logdetLadY - traceMatrix.trace() + m;

		// identity matrix
		MatrixXd Ibig = I_m;

		// backtracing parameters
		double beta = 0.5;

		// initial step size
		double rho1 = rho0;
		vector<int> bbfailures;

		cout << left;
		cout << setw(dispW) << "Rho_BB" << setw(dispW) << "rho" << setw(dispW) << "eps_prim"
			<< setw(dispW) << "res_prim" << setw(dispW) << "eps_dual" << setw(dispW) << "abs(eta)"
			<< setw(dispW) << "iter" << endl;

		// Timer should go here


		// Use AMA to solve the gamma parametized problem
		int AMAstep = 0;
		for (AMAstep = 1; AMAstep <= MaxIter; AMAstep++)
		{
			MatrixXd solveTemp = A.transpose() * Y1 + Y1 * A + C.transpose() * E.cwiseProduct(Y2) * C;
			Xnew = solveTemp.householderQr().solve(Ibig);
			Xnew = (Xnew + Xnew.transpose()) / 2;

			VectorXd eigX = Xnew.eigenvalues().real();
			double logdetX = eigX.log().sum();

			// Gradient of the dual function
			MatrixXd gradD1 = A * Xnew + Xnew * A.transpose();
			MatrixXd gradD2 = E.cwiseProduct(C * Xnew * C.transpose()) - G;

			MatrixXd Rnew2 = gradD2;

			double rho = rho1;

			// declaraitions moved outside for loop so they can be called later

			VectorXd Svecnew;
			complex<double> dualYnew;
			MatrixXd Rnew1;

			for (int j = 1; j <= 50; j++)
			{
				double a = gamma / rho;

				// Z minimization step
				MatrixXd Wnew = -(A * Xnew + Xnew * A.transpose() + (1 / rho) * Y1);
				MatrixXd U; VectorXd s; MatrixXd V;		// declare svd variables
				BDCSVD<MatrixXd> svd(Wnew, ComputeFullV | ComputeFullU);
				U = svd.matrixU();
				V = svd.matrixV();
				s = svd.singularValues();
				VectorXd Svec = s;


				// singular value thresholding
				VectorXd absSvec = Svec.cwiseAbs();
				VectorXd invSvec = absSvec.cwiseInverse();
				VectorXd multSvec = -(a * invSvec);
				multSvec += VectorXd::Ones(multSvec.rows());

				ArrayXd comparison = (Svec.cwiseAbs().array() > a).cast<double>();
				Svecnew = (multSvec.cwiseProduct(Svec)).cwiseProduct(comparison.matrix());

				// update Z
				Znew = U * Svecnew.diagonal() * V.transpose();
				Znew = (Znew + Znew.transpose()) / 2;

				// update Y
				Rnew1 = gradD1 + Znew;

				// Lagrange mulitplier update step
				Y1new = Y1 + rho * Rnew1;
				Y2new = Y2 + rho * Rnew2;
				Y1new = (Y1new + Y1new.transpose()) / 2;
				Y2new = (Y2new + Y2new.transpose()) / 2;

				// e-values of Xinv
				MatrixXd ladYnew = (A.transpose() * Y1new) + (Y1new * A) + (C.transpose() * E.cwiseProduct(Y2new) * C);
				VectorXd eigladYnew = ladYnew.eigenvalues().real();

				complex<double> logdetLadYnew = log(ladYnew.determinant());
				double doubleOfM = m;
				dualYnew = logdetLadYnew - (G * Y2new).trace() + doubleOfM;


				if (eigladYnew.minCoeff() < 0)
				{
					rho = rho * beta;
				}

				else if (dualYnew.real() < dualY.real() +
					(gradD1 * (Y1new - Y1)).trace() +
					(gradD2 * (Y2new - Y2)).trace() -
					(0.5 / rho) * pow((Y1new - Y1).norm(), 2) -
					(0.5 / rho) * pow((Y2new - Y2).norm(), 2))
				{
					rho = rho * beta;
				}

				else
					break;
			}

			// Quinn's code goes here
			// primal residual

			Rnew1 = (A * Xnew) + (Xnew * A.transpose()) + Znew;
			Rnew2 = E.cwiseProduct(C * Xnew * C.transpose()) - G;
			double normRnew1 = Rnew1.norm();
			normRnew1 = pow(normRnew1, 2);
			double normRnew2 = Rnew2.norm();
			normRnew2 = pow(normRnew2, 2);
			double sumNorms = normRnew1 + normRnew2;
			double res_prim = sqrt(sumNorms);


			// calculating the duality gap
			complex<double> eta = -logdetX + (gamma * Svecnew.sum() - dualYnew);

			if (AMAstep % 100 == 0)
			{
				cout << setw(dispW) << rho1 << setw(dispW) << rho << setw(dispW) << eps_prim
					<< setw(dispW) << res_prim << setw(dispW) << eps_dual << setw(dispW) << abs(eta)
					<< setw(dispW) << AMAstep << endl;
			}

			// BB stepsize selection starts here
			MatrixXd Xnew1 = (A.transpose() * Y1new) + (Y1new * A) + (C.transpose() * E.cwiseProduct(Y2new) * C);
			Xnew1 = Xnew1.householderQr().solve(Ibig);
			Xnew1 = (Xnew1 + Xnew1.transpose()) / 2;

			// gradient of dual function
			MatrixXd gradD1new = (A * Xnew1) + (Xnew1 * A.transpose());
			MatrixXd gradD2new = E.cwiseProduct(C * Xnew1 * C.transpose()) - G;

			// numerator for rho1
			MatrixXd changeY1 = Y1new - Y1;
			complex<double> changeY1Norm = changeY1.norm();
			changeY1Norm = pow(changeY1Norm, 2);
			MatrixXd changeY2 = Y2new - Y2;
			complex<double> changeY2Norm = changeY2.norm();
			changeY2Norm = pow(changeY2Norm, 2);
			complex<double> rho1num = changeY1Norm + changeY2Norm;

			// denominator for rho1
			MatrixXd changeGradD1 = gradD1 - gradD1new;
			MatrixXd prodChangY1ChangeGradD1 = changeY1 * changeGradD1;
			complex<double> traceY1 = prodChangY1ChangeGradD1.trace();
			MatrixXd changeGradD2 = gradD2 - gradD2new;
			MatrixXd prodChangY2ChangeGradD2 = changeY2 * changeGradD2;
			complex<double> traceY2 = prodChangY2ChangeGradD2.trace();
			complex<double> rho1den = traceY1 + traceY2;

			complex<double> rho1Complex = rho1num / rho1den;
			rho1 = rho1Complex.real();

			if (rho1 < 0 || isnan(rho1))
			{
				rho1 = rho0;
				bbfailures.push_back(AMAstep);
			}
			// end bb stepsize selection

			// record path of relevant primal and dual quantities

			double detXnew = Xnew.determinant();
			double logdetXnew = log(detXnew);
			double neglogdetXnew = -1 * logdetXnew;
			BDCSVD<MatrixXd> svdZ(Znew, ComputeFullV | ComputeFullU);
			VectorXd svdZnew = svdZ.singularValues();
			double nuc_normZnew = svdZnew.sum();
			nuc_normZnew = nuc_normZnew * gamma;

			output.Jp = neglogdetXnew + nuc_normZnew;
			output.Jd = dualYnew.real();
			output.Rp = res_prim;
			output.dg = abs(eta);
			// NEED AN OUTPUT.TIME


			// various stopping criteria
			if (abs(eta) < eps_dual && res_prim < eps_prim)
			{
				cout << "AMA converged to assinged accuracy!!!" << endl;
				cout << "AMA Steps: " << AMAstep << endl;
				cout << setw(dispW) << rho1 << setw(dispW) << rho << setw(dispW) << eps_prim
					 << setw(dispW) << res_prim << setw(dispW) << eps_dual << setw(dispW) << abs(eta)
					 << setw(dispW) << AMAstep << endl;

				break;
			}

			Y1 = Y1new;
			Y2 = Y2new;
			dualY = dualYnew;


		}


		if (AMAstep == MaxIter)
		{
			output.flag = 0;
		}

		else
		{
			output.flag = 1;
		}

		output.X = Xnew;
		output.Z = Znew;
		output.YOne = Y1new;
		output.YTwo = Y2new;
		output.steps = AMAstep;

	}

	return output;
}

File: ccama.m
% MATLAB implementation of AMA and ADMM algorithms for solving the 
% Covariance Completion problem (CC)
% 
% Written by Armin Zare and Mihailo Jovanovic, April 2016
% 
% minimize -logdetX + gamma*|| Z ||_*
% subject to A*X + X*A' + Z = 0
%            E.*X - G = 0
% A, G, E - problem data
% X, Z    - optimization variables 
%
% Syntax:
% 
% output = ccama(A,C,E,G,gamma,n,m,options)
% 
% Inputs: (1) dynamic matrix A
%             matrix of available output correlations G
%             structural identity matrix E
%             output matrix C
%             low-rank parameter gamma
%             number of masses N
% 
%         (2) options
% 
%             options.rho      - initial step-size
%             options.eps_prim - tolerance on primal residual
%             options.eps_dual - tolerance on duality gap
%             options.maxiter  - maximum number of AMA (ADMM) iterations
%             options.Xinit    - feasible initial value for matrix X
%             options.Zinit    - feasible initial value for matrix Z
%             options.Y1init   - dual-feasible initial value for Y1
%             options.Y2init   - dual-feasible initial value for Y2
%             options.method   - method = 1, AMA (default)
%                              - method = 2, ADMM
% 
% Outputs: output - structured array containing
% 
%          output.X     - matrix X resulting from (CC)
%          output.Z     - matrix Z resulting from (CC)
%          output.Y1    - matrix Y1 resulting from (CC)
%          output.Y2    - matrix Y2 resulting from (CC)
%          output.Jp    - primal objective function at each step
%          output.Jd    - dual objective function at each step
%          output.Rp    - primal residual at each step
%          output.dg    - duality gap at each step
%          output.steps - number of steps for solving (CC)
%          output.time  - cumulative solve time (in seconds) per outer iteration
%          output.flag  - flag = 0, iteration counter reaches its max
%                         flag = 1, problem is solved before maxiter steps
%    
% Additional information:
%
% http://www.umn.edu/~mihailo/software/ccama/
% 

function output = ccama(A,C,E,G,gamma,n,m,options)

% Initialization
if nargin < 7
    error('At least six input arguments are required.')
elseif nargin == 7
    options.rho = 10;
    options.eps_prim = 1.e-5;
    options.eps_dual = 1.e-4;
    options.maxiter = 1.e5;
    Xinit = lyap(A,eye(m));
    options.Xinit = Xinit;
    options.Zinit = eye(m);
    Y1init = lyap(A',-Xinit);
    options.Y1init = gamma*Y1init/norm(Y1init,2);
    options.Y2init = eye(n);
    options.method = 1;
elseif nargin > 8
    error('Too many input arguments.')
end

% Data preprocessing
rho0 = options.rho;
eps_prim = options.eps_prim;
eps_dual = options.eps_dual;
MaxIter = options.maxiter;
X = options.Xinit;
Z = options.Zinit;
Y1 = options.Y1init;
Y2 = options.Y2init;

if options.method == 1      % AMA implementation

    eigLadY = real(eig( A'*Y1 + Y1*A + C'*(E.*Y2)*C ));
    logdetLadY = sum(log(eigLadY));
    dualY = logdetLadY - trace(G*Y2) + m;

    % identity matrix
    Ibig = eye(m);
    % backtracking parameters
    beta  = 0.5;
    % initial step size
    rho1 = rho0;
    bbfailures = [];

    fprintf('%s %s %s %s %s %s %s\n------------------------------------------------------------------------------ \n',...
        'rho_BB', '   rho', '       eps_prim','    res_prim', '    eps_dual',...
        '    abs(eta)','    iter');
    
    % start timer
    tic
    
    % Use AMA to solve the gamma-parameterized problem
    for AMAstep = 1 : MaxIter,

        % X-minimization step
        Xnew = (A'*Y1 + Y1*A + C'*(E.*Y2)*C)\Ibig;
        Xnew = (Xnew + Xnew')/2;

        eigX = real(eig(Xnew));
        logdetX = sum(log(eigX));

        % gradient of the dual function
        gradD1 = A*Xnew + Xnew*A';
        gradD2 = E.*(C*Xnew*C') - G;

        Rnew2 = gradD2;

        rho = rho1;

        for j = 1:50,

            a = gamma/rho;

            % Z-minimization step
            Wnew = -(A*Xnew + Xnew*A' + (1/rho)*Y1);

            [U,S,V] = svd(Wnew);
            Svec = diag(S);

            % singular value thresholding
            Svecnew = ( (1 - a ./ abs(Svec)) .* Svec ) .* (abs(Svec) > a);
            % update Z
            Znew = U*diag(Svecnew)*V';
            Znew = (Znew + Znew')/2;

            % Y- update
            Rnew1 = gradD1 + Znew;

            % Lagrange multiplier update step
            Y1new = Y1 + rho*Rnew1;
            Y2new = Y2 + rho*Rnew2;
            Y1new = (Y1new + Y1new')/2;
            Y2new = (Y2new + Y2new')/2;

            % e-values of Xinv
            eigLadYnew = real(eig( A'*Y1new + Y1new*A + C'*(E.*Y2new)*C ));

            logdetLadYnew = sum(log(eigLadYnew));
            dualYnew = logdetLadYnew - trace(G*Y2new) + m;

            if min(eigLadYnew) < 0             
                rho = rho*beta; 
            elseif (dualYnew < dualY + ...
                        trace(gradD1*(Y1new - Y1)) + ...
                        trace(gradD2*(Y2new - Y2)) - ...
                        (0.5/rho)*norm(Y1new - Y1,'fro')^2 - ...
                        (0.5/rho)*norm(Y2new - Y2,'fro')^2)

                    rho = rho*beta;
            else
                break;   
            end
        end

        % primal residual
        Rnew1 = A*Xnew + Xnew*A' + Znew; 
        Rnew2 = E.*(C*Xnew*C') - G;
        res_prim = sqrt(norm(Rnew1,'fro')^2 + norm(Rnew2,'fro')^2);

        % calculating the duality gap
        eta = -logdetX + gamma*sum(Svecnew) - dualYnew;

        if mod(AMAstep,100) == 0
            disp([num2str(rho1,'%6.1E'),'   ', ...
            num2str(rho,'%6.1E'),'    ', ...
            num2str(eps_prim,'%6.1E'),'      ' ...
            num2str(res_prim,'%6.1E'),'      ' ...
            num2str(eps_dual,'%6.1E'),'      ' ...
            num2str(abs(eta),'%6.1E'),'      ', ...
            num2str(AMAstep,'%i')])
        end

        % BB stepsize selection starts here ...
        Xnew1 = (A'*Y1new + Y1new*A + C'*(E.*Y2new)*C)\Ibig;
        Xnew1 = (Xnew1 + Xnew1')/2;
        % gradient of dual function
        gradD1new = A*Xnew1 + Xnew1*A';
        gradD2new = E.*(C*Xnew1*C') - G;
        rho1 = real(( norm(Y1new - Y1,'fro')^2 + norm(Y2new - Y2,'fro')^2)/ ...
                (trace((Y1new - Y1)*(gradD1-gradD1new)) + trace((Y2new - Y2)*(gradD2-gradD2new))));
        if rho1 < 0 || isnan(rho1)
            rho1 = rho0;
            bbfailures(end+1) = AMAstep;
        end
        % ... and ends here     

        % record path of relevant primal and dual quantities
        output.Jp(AMAstep) = -log(det(Xnew)) + gamma*sum(svd(Znew));
        output.Jd(AMAstep) = real(dualYnew);
        output.Rp(AMAstep) = res_prim;
        output.dg(AMAstep) = abs(eta);
        output.time(AMAstep) = toc;
        
        % various stopping criteria
        if (abs(eta) < eps_dual && res_prim < eps_prim)
            disp('AMA converged to assigned accuracy!!!')
            disp(strcat('AMA steps: ',num2str(AMAstep)))
            disp([num2str(rho1,'%6.1E'),'   ', ...
            num2str(rho,'%6.1E'),'    ', ...
            num2str(eps_prim,'%6.1E'),'      ' ...
            num2str(res_prim,'%6.1E'),'      ' ...
            num2str(eps_dual,'%6.1E'),'      ' ...
            num2str(abs(eta),'%6.1E'),'      ', ...
            num2str(AMAstep,'%i')])
            break;
        end

        Y1 = Y1new;
        Y2 = Y2new;
        dualY = dualYnew;

    end

    if AMAstep == MaxIter
        output.flag = 0;
    else
        output.flag = 1;
    end
    output.X = Xnew;
    output.Z = Znew;
    output.Y1 = Y1new;
    output.Y2 = Y2new;
    output.steps = AMAstep;

elseif options.method == 2      % ADMM implementation
    
    % power iteration algorithm - choose initial condition
    V = rand(m,m);
    V = (V + V')/2;
    eigVmin = min(real(eig(V)));
    if eigVmin < 0,  
        V = 2*abs(eigVmin)*eye(m) + V;  
    end
    V = V/norm(V,'fro');
    % number of iterations
    maxit = 100;
    % size of eigMax
    eigMax = zeros(maxit,1);
    % power iteration starts here ...
    for ind = 1:maxit,
        Y1p = A*V + V*A'; 
        Y2p = E.*(C*V*C');
        Vnew = A'*Y1p + Y1p*A + C'*Y2p*C;
        eigMax(ind) = trace(V*Vnew);
        V = Vnew/norm(Vnew,'fro');
    end
    clear Y1p Y2p V
    % ... and ends here
    
    % parameter mu
    rho = rho0;
    mu = 2*rho*eigMax(end);
    a = gamma/rho;

    fprintf('%s %s %s %s %s %s\n----------------------------------------------------------------------------- \n',...
        'rho','       eps_prim','       res_prim', '       eps_dual',...
        '       abs(eta)','    iter');

    % start timer
    tic
    
    % Use ADMM to solve the gamma-parameterized problem
    for ADMMstep = 1 : MaxIter,

        % X-minimization step
        U1 = -(Z + (1/rho)*Y1);
        U2 = G - (1/rho)*Y2;

        for indX = 1:20

            AadAX = A'*(A*X + X*A') + (A*X + X*A')*A + C'*(E.*(C*X*C'))*C;
            RHS = mu*X - rho*AadAX + rho*(A'*U1 + U1*A + C'*(E.*U2)*C);

            % X - obtained from e-value decomposition of RHS
            [Vrhs,Drhs] = eig(RHS);
            Lvec = diag(Drhs);
            xvec = Lvec/(2*mu) + sqrt( (Lvec/(2*mu)).^2 + 1/mu );

            Xnew = Vrhs*diag(xvec)/Vrhs;
            Xnew = (Xnew + Xnew')/2;
            X = Xnew;

        end

        % Z-minimization step
        Wnew = -(A*Xnew + Xnew*A' + (1/rho)*Y1);
        % svd of Wnew
        [U,S,V] = svd(Wnew,0);
        Svec = diag(S);

        % singular value thresholding
        Svecnew = ( (1 - a ./ abs(Svec)) .* Svec ) .* (abs(Svec) > a);
        % update Z 
        Znew = U*diag(Svecnew)*V';
        Znew = (Znew + Znew')/2;

        % Primal and dual residuals
        Rnew1 = A*Xnew + Xnew*A' + Znew; 
        Rnew2 = E.*(C*Xnew*C') - G;
        res_prim = sqrt(norm(Rnew1,'fro')^2 + norm(Rnew2,'fro')^2);
        res_dual = rho*norm(A'*(Znew-Z)+(Znew-Z)*A,'fro');

        % Lagrange multiplier update step
        Y1new = Y1 + rho*Rnew1;
        Y2new = Y2 + rho*Rnew2;

        % calculating the duality gap
        eigLadYnew = real(eig( A'*Y1new + Y1new*A + C'*(E.*Y2new)*C ));
        logdetLadYnew = sum(log(eigLadYnew));
        dualYnew = logdetLadYnew - trace(G*Y2new) + m;
        eigX = real(eig(Xnew));
        logdetX = sum(log(eigX));
        eta = -logdetX + gamma*sum(Svecnew) - dualYnew;

        if mod(ADMMstep,100) == 0
            disp([num2str(rho,'%6.1E'),'    ', ...
            num2str(eps_prim,'%6.4E'),'      ' ...
            num2str(res_prim,'%6.4E'),'      ' ...
            num2str(eps_dual,'%6.4E'),'      ' ...
            num2str(abs(eta),'%6.1E'),'      ', ...
            num2str(ADMMstep,'%i')])
        end

        % record path of relevant primal and dual quantities
        output.Jp(ADMMstep) = -log(det(Xnew)) + gamma*norm_nuc(Znew);
        output.Jd(ADMMstep) = real(dualYnew);
        output.Rp(ADMMstep) = res_prim;
        output.dg(ADMMstep) = abs(eta);
        output.time(ADMMstep) = toc;
        
        % Stopping criteria
        if (abs(eta) < eps_dual && res_prim < eps_prim)
            disp('ADMM converged to assigned accuracy!!!')
            disp(strcat('ADMM steps: ',num2str(ADMMstep)))
            disp([num2str(rho,'%6.1E'),'    ', ...
            num2str(eps_prim,'%6.4E'),'      ' ...
            num2str(res_prim,'%6.4E'),'      ' ...
            num2str(eps_dual,'%6.4E'),'      ' ...
            num2str(abs(eta),'%6.1E'),'      ', ...
            num2str(ADMMstep,'%i')])
            break;
        end

        X = Xnew;
        Z = Znew;
        Y1 = Y1new;
        Y2 = Y2new;

        % adjust rho to balance the primal and dual residuals
        a1 = 10;
        t1 = 2;
        if res_prim > a1*res_dual
            rho = rho * t1;
            a = gamma/rho;
        elseif res_dual > a1*res_prim
            rho = rho / t1;
            a = gamma/rho;     
        end

    end

    if ADMMstep == MaxIter
        output.flag = 0;
    else
        output.flag = 1;
    end
    output.X = Xnew;
    output.Z = Znew;
    output.Y1 = Y1new;
    output.Y2 = Y2new;
    output.steps = ADMMstep;

end
